core/retry.go

```diff
@@ -0,0 +1,141 @@
+	f := func() (T, error) {
+		var (
+			val T
+			err error
+		)
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					if panic_err, ok := r.(error); ok {
+						err = TransientError{panic_err}
+					} else {
+						err = TransientError{fmt.Errorf("panicked: %v", panic_err)}
+					}
+				}
+			}()
+			val, err = functionToRetry()
+			// Convert the returned `PermanentError` (our implementation) to `backoff.PermanentError`.
+			//This exits the retry loop in the `backoff` library.
+			if perm, ok := err.(PermanentError); err != nil && ok {
+				err = backoff.Permanent(perm.Inner)
+			}
+		}()
+		return val, err
+	}
+
+	randomOption := backoff.WithRandomizationFactor(0)
+
+	initialRetryOption := backoff.WithInitialInterval(time.Millisecond * time.Duration(minDelay))
+	multiplierOption := backoff.WithMultiplier(factor)
+	maxIntervalOption := backoff.WithMaxInterval(time.Millisecond * time.Duration(maxInterval))
+	maxElapsedTimeOption := backoff.WithMaxElapsedTime(time.Millisecond * time.Duration(maxElapsedTime))
+	expBackoff := backoff.NewExponentialBackOff(randomOption, multiplierOption, initialRetryOption, maxIntervalOption, maxElapsedTimeOption)
+	var maxRetriesBackoff backoff.BackOff
+
+	if maxTries > 0 {
+		maxRetriesBackoff = backoff.WithMaxRetries(expBackoff, maxTries)
+	} else {
+		maxRetriesBackoff = expBackoff
+	}
+
+	return backoff.RetryWithData(f, maxRetriesBackoff)
+}
+
+// Retries a given function in an exponential backoff manner.
+// It will retry calling the function while it returns an error, until the max retries.
+// If maxTries == 0 then the retry function will run indefinitely until success
+// from the configuration are reached, or until a `PermanentError` is returned.
+// The function to be retried should return `PermanentError` when the condition for stop retrying
+// is met.
+func Retry(functionToRetry func() error, minDelay uint64, factor float64, maxTries uint64, maxInterval uint64, maxElapsedTime uint64) error {
+	f := func() error {
+		var err error
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					if panic_err, ok := r.(error); ok {
+						err = TransientError{panic_err}
+					} else {
+						err = TransientError{fmt.Errorf("panicked: %v", panic_err)}
+					}
+				}
+			}()
+			err = functionToRetry()
+			// Convert the returned `PermanentError` (our implementation) to a `backoff.PermanentError`.
+			//This exits the retry loop in the `backoff` library.
+			if perm, ok := err.(PermanentError); err != nil && ok {
+				err = backoff.Permanent(perm.Inner)
+			}
+		}()
+		return err
+	}
+
+	randomOption := backoff.WithRandomizationFactor(0)
+
+	initialRetryOption := backoff.WithInitialInterval(time.Millisecond * time.Duration(minDelay))
+	multiplierOption := backoff.WithMultiplier(factor)
+	maxIntervalOption := backoff.WithMaxInterval(time.Millisecond * time.Duration(maxInterval))
+	maxElapsedTimeOption := backoff.WithMaxElapsedTime(time.Millisecond * time.Duration(maxElapsedTime))
```

Comment:
You shouldn't enforce the time to be passed in `time.Milisecond`. `maxElapsedTime` should be of type `time.duration`.

Reply:
addressed